src/main/java/org/jenkinsci/plugins/codesonar/services/ProceduresService.java

```diff
@@ -1,37 +1,101 @@
+import org.jenkinsci.plugins.codesonar.CodeSonarPluginException;
+import org.jenkinsci.plugins.codesonar.CodeSonarRequestURISyntaxException;
+import org.jenkinsci.plugins.codesonar.models.ProcedureMetric;
+import org.jenkinsci.plugins.codesonar.models.json.SearchConfigData;
 import org.jenkinsci.plugins.codesonar.models.procedures.Procedures;
 
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
 /**
  *
  * @author Andrius
  */
-public class ProceduresService {
+public class ProceduresService extends AbstractService {
     private static final Logger LOGGER = Logger.getLogger(ProceduresService.class.getName());
     
     final private HttpService httpService;
     final private XmlSerializationService xmlSerializationService;
+    private boolean strictQueryParameters;
 
-    public ProceduresService(HttpService httpService, XmlSerializationService xmlSerializationService) {
+    public ProceduresService(HttpService httpService, XmlSerializationService xmlSerializationService, boolean strictQueryParameters) {
         this.httpService = httpService;
         this.xmlSerializationService = xmlSerializationService;
+        this.strictQueryParameters = strictQueryParameters;
     }
 
     public URI getProceduresUriFromAnAnalysisId(URI baseHubUri, String analysisId) {
         return baseHubUri.resolve(String.format("/analysis/%s-procedures.xml", analysisId));
     }
     
-    public Procedures getProceduresFromUri(URI proceduresUri) throws IOException {
+    public Procedures getProceduresFromUri(URI proceduresUri) throws CodeSonarPluginException {
         LOGGER.log(Level.INFO, String.format("Calling getProceduresFromUri"));
         
-        InputStream xmlContent = httpService.getContentFromUrlAsInputStream(proceduresUri);
+        HttpServiceResponse response = httpService.getResponseFromUrl(proceduresUri);
+        
+        if(response.getStatusCode() != 200) {
+            throw new CodeSonarHubCommunicationException(proceduresUri, response.getStatusCode(), response.getReasonPhrase(), readResponseContent(response, proceduresUri));
+        }
+
+        return xmlSerializationService.deserialize(response.getContentInputStream(), Procedures.class);
+    }
+    
+    public ProcedureMetric getProcedureWithMaxCyclomaticComplexity(URI baseHubUri, long analysisId) throws CodeSonarPluginException {
+        //Configure search parameters in order to extract the desired data
+        SearchConfigData searchConfig = new SearchConfigData();
+        /*
+         * Read a maximum of 100 procedures with potentially the same cyclomatic complexity value,
+         * sort them client-side by procedure too, and pick-up the very top element in the list.
+         */
+        searchConfig.setLimit(100);
+        searchConfig.getColumns().add("metricCyclomaticComplexity");
+        searchConfig.getColumns().add("procedure");
+        searchConfig.addOrderByCondition("metricCyclomaticComplexity", SearchConfigData.SortingOrder.DESCENDING);
```

Comment:
We should make it clear that we are implementing a work-around for a specific limitation of the Hub API.  I suggest:

```
// We want to find a unique procedure with maximal cyclomatic complexity.
// We do this by searching for procedures and ordering by cyclomatic complexity in descending order.
// Since there may be many procedures all sharing the maximal cyclomatic complexity value,
// we want to further sort by procedure name in order to produce a deterministic result between different analyses.
// Currently in CodeSonar 7.3, the hub API allows us to sort search results on a single field only.
// This means that we must do some of the sorting ourself.
// We don't know how many rows have the maximal cyclomatic complexity,
// so in principle, we must fetch all rows with the maximum and sort them by procedure name.
// Since the number of rows is theoretically too large to fit in memory, doing such a sort is expensive to implement here.
// As a work-around, we fetch 100 rows only, under the assumption that it is unlikely that more than 100 procedures all have the same, maximal cyclomatic complexity.
// If this assumption is wrong, then the results will not be deterministic between analyses,
// but the critical data is the numeric value of the maximal cyclomatic complexity, not the procedure name,
// so we are willing to live with this deficiency.
// Sorting procedures by name will happen inside the MaxCyclomaticComplexityJsonReader.
```

FWIW: I used line comments `//` and not block comments above since one can use a block comment to comment-out a large region of text containing line comments, but one cannot do so if the block contains another block comment.  I don't really care which comment style you ultimately use.